src/jdk.incubator.foreign/share/classes/jdk/internal/foreign/AbstractMemorySegmentImpl.java

```diff
@@ -406,4 +377,52 @@ public int characteristics() {
             return NONNULL | SUBSIZED | SIZED | IMMUTABLE | ORDERED;
         }
     }
+
+    // Object methods
+
+    @Override
+    public String toString() {
+        return "MemorySegment{ id=0x" + Long.toHexString(id()) + " limit: " + length + " }";
+    }
+
+    public static AbstractMemorySegmentImpl ofBuffer(ByteBuffer bb) {
+        long bbAddress = nioAccess.getBufferAddress(bb);
+        Object base = nioAccess.getBufferBase(bb);
+        UnmapperProxy unmapper = nioAccess.unmapper(bb);
+
+        int pos = bb.position();
+        int limit = bb.limit();
+
+        MemoryScope bufferScope = new MemoryScope(bb, null);
+        int size = limit - pos;
+        if (base != null) {
+            return new HeapMemorySegmentImpl<>(bbAddress + pos, () -> (byte[])base, size, defaultAccessModes(size), Thread.currentThread(), bufferScope);
```

Comment:
Seems to be pre-existing, but using the default access modes here would allow dropping access modes from a segment by going to ByteBuffer and then back to MemorySegment. I think at least the read-only property of the BB should be respected here, but maybe we can just retrieve the originally captured MemorySegmentProxy and use the access modes from that?